src/main/java/edu/byu/cs/autograder/git/GitHelper.java

```diff
@@ -34,19 +34,47 @@
 public class GitHelper {
     private static final Logger LOGGER = LoggerFactory.getLogger(GitHelper.class);
     private final GradingContext gradingContext;
+    private String headHash;
 
     public GitHelper(GradingContext gradingContext) {
         this.gradingContext = gradingContext;
     }
 
     // ## Entry Point ##
-    public CommitVerificationResult setUp() throws GradingException {
+    public CommitVerificationResult setUpAndVerifyHistory() throws GradingException {
+        setUp();
+        return verifyCommitHistory();
+    }
+    public void setUp() throws GradingException {
         File stageRepo = gradingContext.stageRepo();
-        fetchRepo(stageRepo);
+        fetchRepo(gradingContext.stageRepo());
+        headHash = getHeadHash(stageRepo);
+    }
+    public CommitVerificationResult verifyCommitHistory() {
+        if (headHash == null) {
+            throw new RuntimeException("Cannot verifyCommitHistory before headHash has been populated. Call setUp() first.");
+        }
```

Comment:
Is it possible headHash could be null for any other reason? Would you be ok making this a GradingException instead of a RuntimeException?

Reply:
## Head Hash nullability
That's a good question about the head hash. 

Head hash is already defined as a `NOT NULL` field in the SQL database, and is required in many areas of the application.

In practice, the head has really can't ever be `null`. A git repo always has a head. This check is mostly verifying that the `setUp()` method has been called before attempting to run the method. I'm providing the developer with a prettier error message if they forget to call the set up the object first. Currently, the `setUp()` method is not ever called from outside the class because the `Grader` now calls `CommitVerificationResult commitVerificationResult = gitHelper.setUpAndVerifyHistory();`

See the create table statement for reference.
```sql
 createSubmissionTableStatement.executeUpdate("""
                        CREATE TABLE IF NOT EXISTS `submission` (
                                `id` INT NOT NULL AUTO_INCREMENT,
                                `net_id` VARCHAR(20) NOT NULL,
                                `repo_url` VARCHAR(200) NOT NULL,
                                `head_hash` VARCHAR(40) NOT NULL,
                                `timestamp` DATETIME NOT NULL,
                                `phase` VARCHAR(9) NOT NULL,
                                `passed` BOOL NOT NULL,
                                `score` FLOAT NOT NULL,
                                `notes` TEXT,
                                `rubric` JSON,
                                `verified_status` VARCHAR(30),
                                `verification` JSON,
                                `admin` BOOL NOT NULL,
                                PRIMARY KEY (`id`),
                                INDEX sort_index (timestamp),
                                CONSTRAINT `net_id`
                                    FOREIGN KEY (`net_id`)
                                    REFERENCES `user` (`net_id`)
                                    ON DELETE CASCADE
                                    ON UPDATE CASCADE
                            ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
                        """);
```

## Throwing a Runtime Exception
One of the points of this newly sub-divided method was to not throw any checked exception. I wanted this to catch all errors inside it, notify the observer, and then continue with the grading. This allows a submission to proceed and be logged in the database as opposed to throwing an error and halting early like happened earlier today.

I'm not sure how a student could end up with a repo without a head hash. In this case, the error isn't checking for student setup issues, but this is meant to report to a developer if create the preconditions properly. A checked exception isn't the right kind of exception to throw in that case, I want this to fail fast during development if a developer does something wrong. In this particular case, it's never something that a developer would be able to recover from.

In the imaginary situation where a student does end up without a head hash, that would be a fatal error anyways. We can't checkout code, grade anything without it. A `RuntimeException` that will halt the grading is appropriate.